Adds in a merge_dup_episode_categories mgmt command

fredkingham opened the pull request:
A patient should never have multiple episodes with the same episode category.

This was not always the case. This PR merges the duplicate episode categories. The logic is mostly taken from merge_patient but does not populate previousMRN fields.

If you're happy with this, after it has been merged we should put in a database constraint, althought that might be slightly tricky as the model is in opal.

There's no reason this is not in 108. The only reason was because it is required by code going in the cerner merge branch so if you want to put it in 108, that works too.

Commit:
Adds in a merge_dup_episode_categories mgmt command

A patient should never have multiple episodes with the same episode
category.

This was not always the case. This PR merges the duplicate episode
categories. The logic is mostly taken from merge_patient but does not
populate previousMRN fields

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+                patient_id = category_episode.patient_id
+                category_episode.delete()
+                with open(DELETE_FILE, 'a') as w:
+                    w.write(f'\n{patient_id},{category_episode_id}')
```
Same as merge patient, can this be a JSON serialisation of to_dict

fredkingham replied:
What do you want to_dicted? Patient or episode?

I've changed it do to episode.to_dict.

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+        if len(category_episodes) > 1:
+            root = category_episodes[0]
+            for category_episode in category_episodes[1:]:
+                merge_patient.update_tagging(category_episode, root)
```
If this is a public API can we have named arguments that let me know from the calling context which way round the data is going to be saved without having to look in the file...
That's prob. another PR
Having looked, I think the comments and docstring on this function both have typos / wrong words

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+            old_subrecord.save()
+
+
+def move_record(subrecord_cls, old_parent, new_parent):
```
Not sure this layer of indirection adds anything

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+
+def move_non_singletons(subrecord_cls, old_parent, new_parent):
+    """
+    Moves the old_subrecords query set onto the new parent (a patient or episode).
```
Can you really move  a queryset?

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+            old_singleton.save()
+    else:
+        if new_singleton.updated < old_singleton.updated:
+            # the old singleton is new than the new singleton
```
*newer

davidmiller commented on intrahospital_api/management/commands/merge_dup_episode_categories.py:
```diff
@@ -0,0 +1,131 @@
+    else:
+        if new_singleton.updated < old_singleton.updated:
+            # the old singleton is new than the new singleton
+            # stamp the new singleton as reversion
```
what is 'stamp' a metaphor for here? not sure I understand this

Commit:
Removes a layer of indirection in merge_dup_episode_categories

Commit:
Renames variables in merge_dup_episode_categories

references to old_parent/new_parent aren't helpful when the parent is
always and episode, lets call a spade a spade.

Commit:
Changes merge_dup_episodes logging

Rather than logging patient id and episode id, log the full episode of
the old episode

fredkingham commented:
You note update_tagging is now a public api as such I have changed it to take keyword arguments and improved the comment here https://github.com/openhealthcare/elcid-rfh/pull/1468

IMO we will ideally add a database constraint to stop the creation of duplicate episode categories so in theory this will hopefully not be a public api for long and we will be able to delete this command.

davidmiller commented:
> You note update_tagging is now a public api as such I have changed it to take keyword arguments and improved the comment here #1468

As is, won't this command error on L102?

That's merged, so I imagine the sequence is 108 -> Cerner Merge, Cerner Merge -> this branch

fredkingham commented:
> > You note update_tagging is now a public api as such I have changed it to take keyword arguments and improved the comment here #1468
> 
> As is, won't this command error on L102?
> 
> That's merged, so I imagine the sequence is 108 -> Cerner Merge, Cerner Merge -> this branch

So I merged cerner merge branch into it. I'm not sure why you think it'll error though.

fredkingham commented:
(added needs discussion as its very possible I'm missing the point of your message)
